server/persistence/mongodb/mongodb.go

```diff
@@ -0,0 +1,416 @@
+package mongodb
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/x/bsonx"
+
+	"github.com/joshprzybyszewski/cribbage/jsonutils"
+	"github.com/joshprzybyszewski/cribbage/model"
+	"github.com/joshprzybyszewski/cribbage/server/interaction"
+	"github.com/joshprzybyszewski/cribbage/server/persistence"
+	"github.com/joshprzybyszewski/cribbage/server/persistence/mongodb/mapbson"
+)
+
+const (
+	dbName          string = `cribbage`
+	gamesCol        string = `games`
+	playersCol      string = `players`
+	interactionsCol string = `interactions`
+)
+
+const (
+	latestGameAction int = -1
+)
+
+var _ persistence.DB = (*mongodb)(nil)
+
+type mongodb struct {
+	bsonRegistry *bsoncodec.Registry
+	client       *mongo.Client
+}
+
+func New(uri string) (persistence.DB, error) {
+	if uri == `` {
+		// If we don't know where to connect, use the default localhost URI
+		uri = `mongodb://localhost:27017`
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
+	if err != nil {
+		return nil, err
+	}
+
+	m := mongodb{
+		client:       client,
+		bsonRegistry: mapbson.CustomRegistry(),
+	}
+
+	// needs to match gameList.GameID
+	err = m.setupIndex(ctx, `gameID`, m.gamesCollection().Indexes())
+	if err != nil {
+		return nil, err
+	}
+
+	// needs to match model.Player.ID
+	err = m.setupIndex(ctx, `id`, m.playersCollection().Indexes())
+	if err != nil {
+		return nil, err
+	}
+
+	// needs to match interaction.PlayerMeans.PlayerID
+	err = m.setupIndex(ctx, `playerID`, m.interactionsCollection().Indexes())
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+func (m *mongodb) setupIndex(ctx context.Context, indexKey string, idxs mongo.IndexView) error {
+	listMaxTimeOpt := &options.ListIndexesOptions{}
+	listMaxTimeOpt.SetMaxTime(30 * time.Second)
+	cur, err := idxs.List(ctx, listMaxTimeOpt)
+	if err != nil {
+		return err
+	}
+
+	for cur.Next(context.Background()) {
+		index := bson.D{}
+		err = cur.Decode(&index)
+		if err != nil {
+			return err
+		}
+		for _, i := range index {
+			if key := i.Key; key == `key` {
+				if val, ok := i.Value.(bson.D); ok && val[0].Key == indexKey {
+					// found the desired index, exit
+					return nil
+				}
+			}
+		}
+	}
+
+	keys := bsonx.Doc{{Key: indexKey, Value: bsonx.Int64(int64(1))}}
```

Comment:
what does the `1` do?

Reply:
It means it's creating an "ascending index": https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/